jib-cli/src/main/java/com/google/cloud/tools/jib/cli/buildfile/Layers.java

```diff
@@ -0,0 +1,179 @@
+   * directories here, or allow passing of the file entry information directly to the reproducible
+   * layer builder
+   *
+   * @param buildRoot the directory to resolve relative paths, usually the directory where the build
+   *     config file is located
+   * @param layersSpec a layersSpec containing configuration for all layers
+   * @return a {@link List} of {@link LayerObject} to use as part of a buildplan
+   * @throws IOException if traversing a directory fails
+   */
+  static List<LayerObject> toLayers(Path buildRoot, LayersSpec layersSpec) throws IOException {
+    List<LayerObject> layers = new ArrayList<>();
+
+    FilePropertiesStack filePropertiesStack = new FilePropertiesStack();
+    // base properties
+    layersSpec.getProperties().ifPresent(filePropertiesStack::push);
+
+    for (LayerSpec entry : layersSpec.getEntries()) {
+      // each loop is a new layer
+      if (entry instanceof FileLayerSpec) {
+        FileEntriesLayer.Builder layerBuiler = FileEntriesLayer.builder();
+
+        FileLayerSpec fileLayer = (FileLayerSpec) entry;
+        layerBuiler.setName(fileLayer.getName());
+        // layer properties
+        fileLayer.getProperties().ifPresent(filePropertiesStack::push);
+        for (CopySpec copySpec : ((FileLayerSpec) entry).getFiles()) {
+          // copy spec properties
+          copySpec.getProperties().ifPresent(filePropertiesStack::push);
+
+          // relativize all paths to the buildRoot location
+          Path rawSrc = copySpec.getSrc();
+          Path src = rawSrc.isAbsolute() ? rawSrc : buildRoot.resolve(rawSrc);
+          AbsoluteUnixPath dest = copySpec.getDest();
+
+          if (Files.isRegularFile(src)) { // regular file
+            if (copySpec.getExcludes().size() > 0 || copySpec.getIncludes().size() > 0) {
+              throw new UnsupportedOperationException(
+                  "Cannot apply includes/excludes on single file copy directives.");
+            }
+            AbsoluteUnixPath target = dest;
+            if (copySpec.isDestEndsWithSlash()) {
+              target = copySpec.getDest().resolve(src.getFileName());
+            }
+            layerBuiler.addEntry(
+                src,
+                target,
+                filePropertiesStack.getFilePermissions(),
+                filePropertiesStack.getModificationTime(),
+                filePropertiesStack.getOwnership());
+          } else if (Files.isDirectory(src)) { // directory
+            List<PathMatcher> excludes =
+                copySpec
+                    .getExcludes()
+                    .stream()
+                    .map(
+                        pattern ->
+                            pattern.endsWith("/") || pattern.endsWith("\\")
+                                ? pattern + "**"
+                                : pattern)
+                    .map(exclude -> FileSystems.getDefault().getPathMatcher("glob:" + exclude))
+                    .collect(Collectors.toList());
```

Comment:
Maybe this part can be factored out as `toPathMatchers(List<String> globPatterns)` (or `globPatternsToPathMatchers()`) and reused?

Reply:
yeah, also wondering if there's a way to combine path matchers like spring has for https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/PathMatcher.html#combine-java.lang.String-java.lang.String-